velox/dwio/parquet/reader/PageReader.cpp

```diff
@@ -365,6 +370,42 @@ void PageReader::prepareDictionary(const PageHeader& pageHeader) {
       }
       break;
     }
+    case thrift::Type::INT96: {
+      auto numVeloxBytes = dictionary_.numValues * sizeof(Timestamp);
+      dictionary_.values = AlignedBuffer::allocate<char>(numVeloxBytes, &pool_);
+      auto numBytes = dictionary_.numValues * sizeof(Int96Timestamp);
+      if (pageData_) {
+        memcpy(dictionary_.values->asMutable<char>(), pageData_, numBytes);
+      } else {
+        dwio::common::readBytes(
+            numBytes,
+            inputStream_.get(),
+            dictionary_.values->asMutable<char>(),
+            bufferStart_,
+            bufferEnd_);
+      }
+      // Expand the Parquet type length values to Velox type length.
+      // We start from the end to allow in-place expansion.
+      auto values = dictionary_.values->asMutable<Timestamp>();
+      auto parquetValues = dictionary_.values->asMutable<char>();
+
+      for (auto i = dictionary_.numValues - 1; i >= 0; --i) {
+        // Convert the timestamp into seconds and nanos since the Unix epoch,
+        // 00:00:00.000000 on 1 January 1970.
+        int64_t nanos;
+        memcpy(
+            &nanos,
+            parquetValues + i * sizeof(Int96Timestamp),
+            sizeof(int64_t));
+        int32_t days;
+        memcpy(
+            &days,
+            parquetValues + i * sizeof(Int96Timestamp) + sizeof(int64_t),
+            sizeof(int32_t));
+        values[i] = Timestamp::fromDaysAndNanos(days, nanos);
```

Comment:
You may need to handle the precision truncation like here later: https://github.com/facebookincubator/velox/pull/10019/files#diff-5dbd33739a55b860fbd2d80af6dfcec1c7883b98b644665a2357f275ed3918c0R153

Reply:
Thanks for the pointer. This issue does exist for timestamp Parquet reader. I added a similar test `ParquetTableScanTest.timestampPrecisionMicrosecond` in this PR to reproduce. Since RowReaderOptions is not accessible in PageReader, I'm thinking of below two options to fix this issue.

1. Creates timestamps with nano precision in PageReader and modifies the precision in TimestampColumnReader based on the timestampPrecision_ in RowReaderOptions.
2. Saves the 12 bytes as int128_t in PageReader, and uses `fromDaysAndNanos` in TimestampColumnReader to convert int128_t values to timestamps with the desired precision.

I'd prefer 1 as TimestampColumnReader handles timestamps so the code can be reused for int64-timestamp and int96-timestamp reading. Additionally, the int64-timestamps may be in milli, micro, or nanoseconds; with the second option, TimestampColumnReader may need to handle the unit difference.
@Yuhta Do you have any preference for the fix? Thanks!

Reply:
Thanks for your review. I implemented as 1 proposed to handle the read precision.
https://github.com/facebookincubator/velox/pull/4680/files#diff-ae87451c1577f3b47d2863187de8bf30c7351484d39537419016487cc7b2f71cR40-R69
This precision is controlled by the added hive configs `kParquetReadTimestampUnit` and `kParquetReadTimestampUnitSession`. Below test has been added to test the precision adjustment according to the config.
> TEST_F(ParquetTableScanTest, timestampPrecisionMicrosecond)